lld/ELF/Arch/RISCV.cpp

```diff
@@ -651,12 +651,38 @@ static void relaxHi20Lo12(const InputSection &sec, size_t i, uint64_t loc,
   if (!isInt<12>(r.sym->getVA(r.addend) - gp->getVA()))
     return;
 
+  // The symbol may be accessed in multiple pieces with different addends.
+  // If we are relaxing the HI20 relocation, we need to ensure that we only
+  // relax (and delete the instruction) if all possible LO12 relocations
+  // that depend on it will be relaxable. The compiler will only access multiple
+  // pieces of an object with low relocations on the memory op if the alignment
+  // allows it. Therefore it should suffice to check that the smaller of the
+  // alignment and size can be reached from GP.
+  uint32_t alignAdjust =
+      r.sym->getOutputSection() ? r.sym->getOutputSection()->addralign : 0;
+  alignAdjust = std::min<uint32_t>(alignAdjust, r.sym->getSize());
+  if (alignAdjust)
+    alignAdjust--;
+
   switch (r.type) {
-  case R_RISCV_HI20:
+  case R_RISCV_HI20: {
+    uint64_t hiAddr = r.sym->getVA(r.addend);
+    // If the addend is zero, the LO12 relocations can only be accessing the
+    // range [base, base+alignAdjust] (where base == r.sym->getVA()).
+    if (r.addend == 0 && !isInt<12>(hiAddr + alignAdjust - gp->getVA()))
+      return;
+
+    // However, if the addend is non-zero, the LO12 relocations may be accessing
+    // the range [HI-alignAdjust-1, HI+alignAdjust].
+    if (r.addend != 0 && (!isInt<12>(hiAddr - alignAdjust - 1 - gp->getVA()) ||
```

Comment:
The addend ==0 and !=0 distinction appears to make the condition less strict and utilize pointer provenance: for `char p[1], q[1];`, we can't use q-1 to get p even if q is placed after p.

At the object file format level, I wonder whether we should be less strict.
I understand that being less strict would not be able to relax the first `lui` in `riscv-relax-hi20-lo12.s`.

Reply:
I suppose we could check the negative range even in the `addend==0` case with the assumption that the compiler won't allow for out-of-bounds access as you describe here. This is just another one of those assumptions that the linker makes about what the compiler will or will not do.
There are of course other ways we can accomplish this:
- Disable relaxation if we see a negative addend on a `LO12` relaxation
- Keep track of symbols that are accessed with such negative addends and disable relaxation for relocations referencing those symbols
- Emit a fatal error if such a negative addend is used

Reply:
I want to hear from others about the reliability.

Ultimately I hope that we can derive some rules that are simple (ideally no `addend==0`/`addend!=0` distinction), even if we give up some legitimate opportunities.

Reply:
How about if I just change it to reject relaxation for negative addends? That way, we always look only forward? Would that be too pessimistic?